python/pyiceberg/expressions/literals.py

```diff
@@ -134,55 +134,83 @@ def literal(value: L) -> Literal[L]:
         raise TypeError(f"Invalid literal value: {repr(value)}")
 
 
-class FloatAboveMax(Literal[float], Singleton):
+class AboveMax(Literal[L]):
+    ...
+
+
+class BelowMin(Literal[L]):
+    ...
+
+
+class FloatAboveMax(AboveMax[float], Singleton):
     def __init__(self):
         super().__init__(FloatType.max, float)
 
+    @singledispatchmethod
     def to(self, type_var: IcebergType) -> Literal:  # type: ignore
         raise TypeError("Cannot change the type of FloatAboveMax")
 
+    @to.register(FloatType)
+    def _(self, _: FloatType) -> Literal[float]:
+        return self
+
     def __repr__(self) -> str:
         return "FloatAboveMax()"
 
     def __str__(self) -> str:
         return "FloatAboveMax"
 
 
-class FloatBelowMin(Literal[float], Singleton):
+class FloatBelowMin(BelowMin[float], Singleton):
     def __init__(self):
         super().__init__(FloatType.min, float)
 
+    @singledispatchmethod
     def to(self, type_var: IcebergType) -> Literal:  # type: ignore
         raise TypeError("Cannot change the type of FloatBelowMin")
 
+    @to.register(FloatType)
+    def _(self, _: FloatType) -> Literal[float]:
```

Comment:
If you would supply the correct literal to the unbound predicate:
```
EqualTo('a', literal(IntegerType.max + 1).to(IntegerType()))
```
Then we would convert this again, and raise an exception. I ran into this when writing tests. This is a bit more friendly, and I don't think there is any harm in it to allow this.

Reply:
We generally don't want these to leak outside of expression binding, but I see no harm in making it a bit more friendly.